axlearn/common/test_utils.py

```diff
@@ -230,6 +238,16 @@ def assertNestedEqual(self, a, b):
             if hasattr(a_value, "dtype"):
                 self.assertEqual(a_value.dtype, b_value.dtype)
 
+    def assertTensorAllClose(self, x, y, rtol=1e-5, atol=1e-5, check_dtypes=True, **kwargs):
+        """assert Tensor are all close."""
+        x = np.asarray(x)
+        y = np.asarray(y)
+        if check_dtypes:
+            self.assertDtypesMatch(x, y)
+        x = x.astype(np.float32) if x.dtype == jnp.bfloat16 else x
+        y = y.astype(np.float32) if y.dtype == jnp.bfloat16 else y
+        np.testing.assert_allclose(x, y, rtol=rtol, atol=atol, **kwargs)
+
```

Comment:
@markblee to comment if we want to add this util.

Reply:
Can we use `assertNestedAllClose` directly?

Reply:
+1, I think I left another comment on this as well.

Reply:
Unfortunately not. `assertNestedAllClose` will first `flatten_items`. For example, if we do 

```
self.assertTensorAllClose(jnp.array[0, 0, 0, 0], [0] * 4)
```

it will flatten the first item to 

```
[('', Array([1., 1., 1., 1.], dtype=float32))]
```
and the second item to
```
[('0', 1.0), ('1', 1.0), ('2', 1.0), ('3', 1.0)]
```

Obviously, they are different, so the test will fail. 

I also tried something more verbose: 

```
x = {'tensor': jnp.array([1.0] * 4)}
y = {'tensor': [1.0] * 4}
self.assertNestedAllClose(x, y)
```
Again, `flatten` will expand them differently. 

Another possibility is to use something like 

```
x = jnp.array([1.0] * 4)
y = [1.0] * 4
self.assertNestedAllClose(x, jnp.array(y))
```

This will work, but under the hook, it is expanding the jnp/np array to a list of elements, and compare them one-by-one. Of course, if we do

```
x = jnp.array([1.0] * 4)
y = [1.0] * 4
self.assertNestedAllClose(
    {'tensor': x},
    {'tensor': jnp.array(y)})
```

it will work, but this is too troublesome from my biased perspective. 

It is pretty common to compare a jnp/numpy tensor with an expected tensor or list of elements. I have seen people start to use something like https://github.com/apple/axlearn/blob/1a045eae42bda6fcefbf5f4decbebc1b7969ce78/axlearn/vision/nms_test.py#L38, but that requires to subclass from tf.test.TestCase, which brings even more problems from my point of view.

Reply:
```
x = jnp.array([1.0] * 4)
y = jnp.array([1.0] * 4)
self.assertNestedAllClose(x, y)
```
seems reasonable to me (we treat the Tensors as leaves). Can you elaborate on the concern?

Reply:
First we need to convert python list to jax numpy array, which needs extra work in writing test; second this requires the dtype of both x and y must match, but what we only mean to be compare the numerical value.

Reply:
We can change `assertNestedAllClose` to be more relaxed on dtypes. There are also ways to make it support `[1.0] * 4`. 

We should decouple this change to a separate PR and discuss separately.

Reply:
Thanks for the comment. I have removed the change in `test_utils.py`. will follow up in another cl if necessary.

Reply:
Done.